Tables and HTS service deployed on AKS (#139)
linkedin/openhouse: infra/recipes/terraform/azure/environments/sandbox/README.md

`export TF_VAR_storage_account_name="foobar"`

### Using tfvars
1. Create a copy of the file `backend.tfvars.template` into your environment directory by running
`cp backend.tfvars.template ./environments/<environment-name>/backend.tfvars`.
2. Edit the backend.tfvars to configure the necessary TF backend state.

> ## Docker setup
> 
> To be able to build the Tables and Housetables service images, you will need to ensure Docker is installed and running on your local machine. You can check for this by running a command like `docker info` in the terminal.
> 
> Sometimes, the images may stop building or uploading to the ACR, a process that should be completed automatically after the `terraform apply` commands. If a docker related error is shown, it is likely because docker has run out of local memory. To fix this, run `docker system prune --volumes` locally.
> 
> ### Building OpenHouse project
> 
> Before running `terraform apply`, run `./gradlew clean build` from the root directory to configure the JAR files for image builds.

Reviewer: ```suggestion
## Docker Images - Build

Follow the steps described in [SETUP.md](SETUP.md#build-containers) to build docker container images locally for OpenHouse services. Currently, running only tables and housetables service is supported in Azure.

```

Author: I am not sure this is fully accurate. All the user needs to do here is run `./gradlew clean build`. The `terraform apply` builds the images and publishes them to the ACR.

Reviewer: Got it, that is cool. I updated my recommendation. I didn't fully follow what was supported in container.

`cp backend.tfvars.template backend.tfvars`.
2. Edit the backend.tfvars to configure the necessary TF backend state.

## Docker setup

To be able to build the Tables and Housetables service images, you will need to ensure Docker is running on your local machine. This is typically done by opening Docker Desktop. You can check for this by running a command like `docker info` in the terminal.

Sometimes, the images may stop building or uploading to the ACR. This is likely because docker has run out of local memory. To fix this, run `docker system prune --volumes` locally.

### Building OpenHouse project

Before building the images, run `./gradlew clean build` from the root directory. This only works on Java8.

> ## Deployment

Reviewer 2: Can we update the README with k8s deployment using helm provider?

Author: Yes - what more should I include?

Reviewer 2: Yes mainly on the helm chart integration part as part of terraform and the services that are installed.

Author: Done

`cp backend.tfvars.template backend.tfvars`.
2. Edit the backend.tfvars to configure the necessary TF backend state.

> ## Docker setup

Reviewer: Just so I understand this, README flow should be organized in the order you want to user to execute the commands. After TF Backend setup do you want them to setup Docker or TF plan/apply for the backend?

If you need ACR provisioned, I would there are various ways in which we could organize the flow.

Option 1:
1. TF Backend setup for ACR
2. Docker image building for OH and publish to ACR
3. TF Backend setup for rest of infra (including deploying of services)

Option 2:
1. TF Backend setup for rest of infra (w/ ACR and w/o OH services deployment)
2. Docker image building for OH and publish to ACR
3. TF Backend setup only for OH services deployment.

Either is fine, but whichever we pick README should follow the flow clearly.

> To be able to build the Tables and Housetables service images, you will need to ensure Docker is running on your local machine. This is typically done by opening Docker Desktop. You can check for this by running a command like `docker info` in the terminal.

Reviewer: ```suggestion
To be able to build the Tables and Housetables service images, you will need to ensure Docker is installed on your local machine. 
```

Author: It needs to be both installed and currently running; the image build doesn't work if I don't have the docker desktop open

> Sometimes, the images may stop building or uploading to the ACR. This is likely because docker has run out of local memory. To fix this, run `docker system prune --volumes` locally.

Reviewer: Can you add step by step instruction to build and upload to ACR?

Author: Yes, it is mostly done automatically by running `terraform apply`, but I will include that

> ### Building OpenHouse project
> 
> Before building the images with Terraform, run `./gradlew clean build` from the root directory to configure the JAR files for image builds. This only works on Java8.

Reviewer: ```suggestion
### Build OpenHouse project

Run `./gradlew clean build` from the root directory to build the JAR files that we use in docker image.

### Building Docker image artifacts

<ADD commands to do that>

### Upload Docker image to Azure container registry

<ADD commands to do that>
```

Author: These other two sections are automatically done with `terraform apply`. The user does not need to do any additional commands here.

## Helm releases

> The sandbox uses helm releases to pull the specific service's image from the ACR repository and then deploy it on AKS. This also requires a Kubernetes secret to authenticate with the MySQL database.

Reviewer: This is not super useful section if there is no action expected from the user. Is this meant for documentation fyi?

Author: Yes I included this section in response to Abhishek's comment [here](https://github.com/linkedin/openhouse/pull/139#discussion_r1688990479). I will delete

_If using tfvars_
`terraform init -backend-config="./backend.tfvars"`

_If using environment variables_
`terraform init`

### Planning & Applying

`terraform plan` to see the plan generated by terraform for the specified deployment configuration. If you are satisfied with the plan, run `terraform apply` to apply the configuration.

> You will need to run `terraform init`, `terraform plan` and `terraform apply` from two dirrent directories. First, run the commands from the `../container/` directory to set up the Azure Container Registry. Then, run the commands from the `sandbox/` (current) directory to set up the rest of the sandbox.

Reviewer: ```suggestion
#### Docker Images - Publish

Docker container images that are built locally should be published to Azure Container Registry (ACR). You can use the Terraform code provided in [Azure Container Terraform](infra/recipes/terraform/azure/environments/container) to push the images to ACR.

Run the following three commands for ACR publishing

1. `terraform init` : Run it first time to setup tfbackend
2. `terraform plan` : Plan the Terraform deployment
3. `terraform apply` : Apply the Terraform deployment

If you don't want to use terraform, you can always use `docker push` command to publish the container images available in local docker to ACR.

#### Rest of Infra - Setup

As part of bringing OpenHouse in Azure, you will need to setup other Azure services like AKS, MySQL, Storage. You can use the Terraform code provided in [Azure Sandbox Terraform](infra/recipes/terraform/azure/environments/sandbox) to setup rest of infra that is required for successfully bringing up OpenHouse.

Run the following three commands for rest of infra setup

1. `terraform init` : Run it first time to setup tfbackend
2. `terraform plan` : Plan the Terraform deployment
3. `terraform apply` : Apply the Terraform deployment
```

Reviewer 2: We could add the below main terraform command in a separate section and refer in both the places above.

Author: I will add this but I just want to clarify about the above comment before -- the images do not need to be built manually, only the jar files

Reviewer: I see. Updated accordingly

`terraform plan` to see the plan generated by terraform for the specified deployment configuration. If you are satisfied with the plan, run `terraform apply` to apply the configuration.

> You will need to run `terraform plan` and `terraform apply` from two dirrent directories. First, run the commands from the `../container/` directory to set up the Azure Container Registry. Then, run the commands from the `sandbox/` (current) directory to set up the rest of the sandbox.

Reviewer 3: Just realize that we need to run terraform twice. Is it possible to make it one step?

Author: See my slack message and my [comment](https://github.com/linkedin/openhouse/pull/139#issuecomment-2248820066). As of now I don't see a simple way.

One alternative is to write a separate script that combines them. What do you think?

Reviewer 3: As synced offline, it's ok to have two steps.

## Deconfiguration

Run `terraform destroy` once you are done to tear down the services you have created.

# Access Created Services

Once you have successfully run the above steps so the sandbox has been run, you can now access the pods in AKS in your local terminal by ensuring you are logged in to the Azure CLI with `az login` and then running
`az aks get-credentials --resource-group <resource-group> --name <cluster-name>`. You can then run any `kubectl` commands.

## Make requests to services

> To make requests to the services, set up port forwarding by running `kubectl port-forward <pod-name> 8080:8080`. Then, you can make requests to HTS and Tables service via `http:localhost:8080`.

Reviewer 3: `http://localhost:8080`.
Also, if want to make requests to HTS, we need to forward another port likely 8081.

Author: Okay I will add that
